### PR TITLE
Ruler ignore users

### PR DESCRIPTION
**Pull request by pstibrany**
**What this PR does**: This PR adds `-ruler.enabled-tenants` and `-ruler.disabled-tenants` options to explicitly enable or disable rules processing for specific tenants. Code is extracted from Compactor, where this support already exists.

**Checklist**
- [x] Tests updated
- [x] Documentation added
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

File: docs/configuration/config-file-reference.md [hunk 1]
```diff
@@ -1584,6 +1584,18 @@ ring:
...
+# Comma separated list of tenants whose rules this ruler can evaluate. If
+# specified, only these tenants will be handled by ruler, otherwise this ruler
+# can process rules from all tenants. Subject to sharding.
+# CLI flag: -ruler.enabled-tenants
+[enabled_tenants: <string> | default = ""]
+
+# Comma separated list of tenants whose rules this ruler cannot evaluate. If
+# specified, and ruler would normally pick given tenant for processing (eg. via
+# sharding), tenant will be ignored instead.
+# CLI flag: -ruler.disabled-tenants
```
Comment (gotjosh) [inline docs/configuration/config-file-reference.md:1597]:
```suggestion
# Comma separated list of tenants whose rules this ruler cannot evaluate. If
# specified, a ruler that would normally pick the specified tenant(s) for processing
# will ignore them instead. Subject to sharding. 
# CLI flag: -ruler.disabled-tenants
```

Comment (gotjosh) [reply docs/configuration/config-file-reference.md:1597]: The homogenous wording indicates that this process is the exact opposite of the above. Differences in words here would leave an operator puzzled on whenever this means something different. WDYT? 

Feel free to move this to the where this text is generated.

File: pkg/util/allowed_users.go [hunk 2]
```diff
@@ -0,0 +1,45 @@
+package util
+
+type AllowedUsers struct {
```
Comment (gotjosh) [inline pkg/util/allowed_users.go:3]: How do you feel about renaming this struct (including variables and file names) to `AllowedTenants`? - I think at this point we've settled on the word tenant in preference over users.

**Review by gotjosh — COMMENTED** [review]
Code looks 👌 , I just have a couple of comments on naming and wording.

File: pkg/ruler/ruler.go [hunk 3]
```diff
@@ -250,6 +259,13 @@ func NewRuler(cfg Config, manager MultiTenantManager, reg prometheus.Registerer,
 		}, []string{"reason"}),
 	}
 
+	if len(cfg.EnabledTenants) > 0 {
+		level.Info(ruler.logger).Log("msg", "using enabled users", "enabled", strings.Join(cfg.EnabledTenants, ", "))
```
Comment (pracucci) [inline pkg/ruler/ruler.go:263]: [nit] in single-binary mode may not be cristal clear who is logging it:

```suggestion
		level.Info(ruler.logger).Log("msg", "ruler is using enabled users", "enabled", strings.Join(cfg.EnabledTenants, ", "))
```

File: pkg/ruler/ruler.go [hunk 4]
```diff
@@ -250,6 +259,13 @@ func NewRuler(cfg Config, manager MultiTenantManager, reg prometheus.Registerer,
...
+	}
+	if len(cfg.DisabledTenants) > 0 {
+		level.Info(ruler.logger).Log("msg", "using disabled users", "disabled", strings.Join(cfg.DisabledTenants, ", "))
```
Comment (pracucci) [inline pkg/ruler/ruler.go:266]: [nit] in single-binary mode may not be cristal clear who is logging it:

```suggestion
		level.Info(ruler.logger).Log("msg", "ruler is using disabled users", "disabled", strings.Join(cfg.DisabledTenants, ", "))
```

File: pkg/compactor/compactor.go [hunk 5]
```diff
@@ -321,20 +317,9 @@ func newCompactor(
 	}
 
 	if len(compactorCfg.EnabledTenants) > 0 {
-		c.enabledUsers = map[string]struct{}{}
-		for _, u := range compactorCfg.EnabledTenants {
-			c.enabledUsers[u] = struct{}{}
-		}
-
 		level.Info(c.logger).Log("msg", "using enabled users", "enabled", strings.Join(compactorCfg.EnabledTenants, ", "))
```
Comment (pracucci) [inline pkg/compactor/compactor.go:320]:
```suggestion
		level.Info(c.logger).Log("msg", "compactor is using enabled users", "enabled", strings.Join(compactorCfg.EnabledTenants, ", "))
```

File: pkg/compactor/compactor.go [hunk 6]
```diff
@@ -321,20 +317,9 @@ func newCompactor(
...
 	}
-
 	if len(compactorCfg.DisabledTenants) > 0 {
-		c.disabledUsers = map[string]struct{}{}
-		for _, u := range compactorCfg.DisabledTenants {
-			c.disabledUsers[u] = struct{}{}
-		}
-
 		level.Info(c.logger).Log("msg", "using disabled users", "disabled", strings.Join(compactorCfg.DisabledTenants, ", "))
```
Comment (pracucci) [inline pkg/compactor/compactor.go:323]:
```suggestion
		level.Info(c.logger).Log("msg", "compactor is using disabled users", "disabled", strings.Join(compactorCfg.DisabledTenants, ", "))
```

File: pkg/util/allowed_users.go [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+
+	// If empty, no users are disabled. If not empty, users in the map are disabled.
+	disabled map[string]struct{}
+}
+
+func NewAllowedUsers(enabled []string, disabled []string) *AllowedUsers {
+	a := &AllowedUsers{}
+
+	if len(enabled) > 0 {
+		a.enabled = map[string]struct{}{}
```
Comment (pracucci) [inline pkg/util/allowed_users.go:15]: [nit]

```suggestion
		a.enabled = make(map[string]struct{}, len(enabled))
```

File: pkg/util/allowed_users.go [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+		for _, u := range enabled {
+			a.enabled[u] = struct{}{}
+		}
+	}
+
+	if len(disabled) > 0 {
+		a.disabled = map[string]struct{}{}
```
Comment (pracucci) [inline pkg/util/allowed_users.go:22]: [nit]

```suggestion
		a.disabled = make(map[string]struct{}, len(disabled))
```

**Review by pracucci — APPROVED** [review]
Good job! LGTM! 🚀

File: docs/configuration/config-file-reference.md (hunk 1 above)
Comment (pstibrany) [reply docs/configuration/config-file-reference.md:1597]: I will use your wording. Thanks for suggestion.

File: pkg/ruler/ruler.go (hunk 3 above)
Comment (pstibrany) [reply pkg/ruler/ruler.go:263]: It should be visible from `caller`, but let's fix the messages too. 👍

File: pkg/util/allowed_users.go (hunk 2 above)
Comment (pstibrany) [reply pkg/util/allowed_users.go:3]: SGTM, good idea.

**Comment by pstibrany** [comment]
Thanks for your reviews! I've now addressed all your feedback.

File: pkg/ruler/ruler_test.go [hunk 9]
```diff
@@ -277,6 +277,8 @@ func TestSharding(t *testing.T) {
 		shardingStrategy string
 		shuffleShardSize int
 		setupRing        func(*ring.Desc)
+		enabledUsers     []string
```
Comment (gotjosh) [inline pkg/ruler/ruler_test.go:280]: nit: not sure if intended but we're still using the Users suffix throughout this test.

File: pkg/ruler/ruler.go [hunk 10]
```diff
@@ -224,20 +230,23 @@ type Ruler struct {
 	ringCheckErrors prometheus.Counter
 	rulerSync       *prometheus.CounterVec
 
+	allowedUsers *util.AllowedTenants
```
Comment (gotjosh) [inline pkg/ruler/ruler.go:233]:
```suggestion
	allowedTenants *util.AllowedTenants
```

File: pkg/compactor/compactor.go [hunk 11]
```diff
@@ -178,12 +178,7 @@ type Compactor struct {
 	logger       log.Logger
 	parentLogger log.Logger
 	registerer   prometheus.Registerer
-
-	// If empty, all users are enabled. If not empty, only users in the map are enabled (possibly owned by compactor, also subject to sharding configuration).
-	enabledUsers map[string]struct{}
-
-	// If empty, no users are disabled. If not empty, users in the map are disabled (not owned by this compactor).
-	disabledUsers map[string]struct{}
+	allowedUsers *util.AllowedTenants
```
Comment (gotjosh) [inline pkg/compactor/compactor.go:181]:
```suggestion
	allowedTenants *util.AllowedTenants
```

**Review by gotjosh — APPROVED** [review]
LGTM!

File: pkg/ruler/ruler_test.go (hunk 9 above)
Comment (pstibrany) [reply pkg/ruler/ruler_test.go:280]: I would need to change entire test to replace that. Doesn't seem to be worth the effort. I've changed ruler.go and compactor.go though.

**Review by gotjosh — APPROVED** [review]

**Pull request merged**